src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala

```diff
@@ -178,16 +180,60 @@ object ZipAndJarSourcePathFactory extends ZipAndJarFileLookupFactory {
 
 final class FileBasedCache[T] {
   import java.nio.file.Path
-  private case class Stamp(lastModified: FileTime, fileKey: Object)
+  private case class Stamp(size: Long, signatures: ArrayBuffer[Long], fileKey: Object)
   private val cache = collection.mutable.Map.empty[Seq[Path], (Seq[Stamp], T)]
 
+  /**
+    * Read the CRCs of every zip entry if all of them succeed to be read.
+    *
+    * This method uses `java.util.jar` instead of `java.util.zip` utilities
+    * because the method `getCrc` in [[java.util.zip.ZipEntry]] returns 0 or
+    * -1 values.
+    *
+    * @param jarFile A jar file (which follows the zip specification format).
+    * @return A list of checksums for every zip entry.
+    */
+  def readJarCrcs(jarFile: Path): Option[ArrayBuffer[Long]] = {
+    try {
+      val jf = new java.util.jar.JarFile(jarFile.toFile)
+      try {
+        val entries = jf.entries()
+        val buffer = new ArrayBuffer[Long](32)
+        while (entries.hasMoreElements) buffer.+=(entries.nextElement().getCrc)
+        // It is very likely that a jar with no CRCs entries is ill-generated, assume so
+        if (buffer.isEmpty || buffer.forall(_ == -1L)) None else Some(buffer)
```

Comment:
```
if (buffer.isEmpty || buffer.forall(_ == -1L)) None else Some(buffer)
```
should this be a `forall` or `forany`?
If for example the LICENCE.TXT had a CRC that was valid but all of the .class files were erased CRC this would not be safe IMO

clearly if this was `forany` it should be re-written to be fast-fail